nimbus/src/dbcache.rs

```diff
@@ -41,7 +49,8 @@ impl DatabaseCache {
         let data = CachedData {
             experiment_branches: eb,
         };
-        // then swap it in.
+        // Commit and swap it in.
+        writer.commit()?;
```

Comment:
Is this when we're flushing to disk? If so, can we `replace` the data before that happens?

Reply:
> Is this when we're flushing to disk?

It is, yes.

> If so, can we replace the data before that happens?

Yes, but we'll need to be careful about what happens if the commit itself fails. I'm imagining something like:

```
// Get write access to the cache
let cache = self.data.write().unwrap();
// Swap it in. Nobody can read this yet, because of the lock above.
let old_data = cache.replace(data);
// Try to flush the changes to the db on disk
match writer.commit() {
  // Phew, that worked out!
  Ok(_) => Ok(()),
  // Quick, roll back the cache changes before anyone can notice!
  Err(e) => {
    if old_data.is_some() {
        cache.replace(old_data);
    } else {
        cache.take();
    }
    Err(e)
  }
}
// Exclusive access to `self.data` goes out of scope here.
```

@jhugman does that make sense, and seem worthwhile?

Reply:
Actually I think it can be simpler than that:

* Take the cache lock
* Commit the transaction
* If it succeeds, update the in-memory cache

I think this will have the same effect of preventing other threads from observing the new data on disk while the old data is still in memory.

Reply:
I've gone ahead and made the change here, and added some comments to explain what I think it's doing and why I think it's sound.

Reply:
> 
>   * Take the cache lock
>   * Commit the transaction
>  * If it succeeds, update the in-memory cache

Sorry, I missed this. 

Checking if I understand [`RwLock`][1] sufficiently: reading from the cache from _the main thread is blocked_ while the data is being written to disk?

[1]: https://doc.rust-lang.org/std/sync/struct.RwLock.html

Reply:
> @jhugman does that make sense, and seem worthwhile?

I think so. One of the things I love about this design is we can optimize the caching in a single place it an makes a difference everywhere.

Reply:
> reading from the cache from the main thread is blocked while the data is being written to disk?

From any thread, not just the main thread; calling `self.data.write().unwrap()` takes the write lock and gives us exclusive access to the data it protects, if another thread tries to read the cache then its call to `self.data.read().unwrap()` will block until we release the write lock.

So...I guess the main thread could block waiting for the I/O being performed by `writer.commit()` here. If we want to avoid that, I think fundamentally we'll need to allow the main thread the possibility of reading either stale or uncommitted data from the cache. Thoughts?

Reply:
If we want to avoid the potential for blocking the main thread, my weakly-held preference is to allow the main thread to read stale data from the cache, by reverting to the original ordering of events:

* Commit the changes to the db.
* *[there is potential for other threads to read stale data between these two steps]*
* Take the write lock on the cache.
* *[potential for reading stale data has now ended]*
* Update the cache.
* Release the write lock.

Reply:
In slack, @mhammond talked me into making this a more strongly-held preference. We seem highly unlikely to write any code that would be upset about seeing stale data in the cache here in practice, and such code seems likely to be broken in other ways. So, I've reverted to the logic I had here previously (commit the transaction, then lock and update the cache) but added a comment to explain what's going on and why.

Reply:
> We seem highly unlikely to write any code that would be upset about seeing stale data in the cache here in practice, and such code seems likely to be broken in other ways

I _think_ I disagree with both of these points though admittedly I'm not sure who the "we" is.

My motivators in this discussion are:

0. [Not waiting on I/O on the main thread][1].
1. making up-to-date data available to the main thread as quickly as possible after startup.

I keep talking about the main thread; I appreciate that any thread reading from the cache will  be blocked, but the main thread is where the user perceives it— as dropped frames when drawing the UI and/or an unresponsive UI. The Fenix team is _extremely_ sensitive to this: low-end devices with slow disks are common is their user population. At this stage, the Fenix Perf team sees only performance regressions, not the value that nimbus is bringing. 

Second: by using a read only `DatabaseCache`, we have chosen freeing up the main thread and periods where we what is in the cache does not match what is on disk.

1. **Stale data or no data**. If we write to disk before refreshing the cache, we serve stale data or at startup we don't have a cache at all. Frequent worst case: startup experiments get `DatabaseNotReady`. Of the two: I am more concerned about `DatabaseNotReady` on startup problem than stale data.
2. **Uncommitted data**. If we [refresh the cache before writing to disk][2], we have a period of time where disk errors can happen and we lose the enrollment calculation. Rare worst case: we lose the disk (we can't do anything about) or we re-do the enrollment calculation after the next fetch/apply cycle.

What I don't know is that is this an optimization that is worth doing? At this stage, it doesn't add any additional complexity to the implementation, and it's not a fundamental design decision.

Are there better cheaper mitigations for the "`DatabaseNotReady` at startup problem" that we should take before this one, or should we do this now.

If we chose _not_ to do this, we should probably record this in an issue as a potential optimization.

[1]: https://github.com/mozilla/nimbus-sdk/pull/94/files#r567497717
[2]: https://github.com/mozilla/nimbus-sdk/pull/94/files#r567496810

Reply:
> I'm not sure who the "we" is.

By "we" here I meant "the nimbus-sdk developers"; this cache is an implementation detail, any code that touches it is code we've deliberately written to use the cache for some specific reason (such as the not-doing-IO-on-the-main-thread thing).

> My motivators in this discussion are:
>
> * Not waiting on I/O on the main thread.
> *making up-to-date data available to the main thread as quickly as possible after startup.

I agree with the importance of both of these.

> by using a read only DatabaseCache, we have chosen freeing up the main thread and periods where
> we what is in the cache does not match what is on disk.

I agree that's what we're choosing (and what we *should* choose) but I would not frame this as a "read-only DatabaseCache", we're still updating it, we're just not blocking any consumers while we attempt to update it atomically with disk IO.

> If we write to disk before refreshing the cache, we serve stale data or at startup we don't have a cache at all.

At startup we will *always* have no cache, since it's just an in-memory hashmap. So I don't think this changes the startup behaviour.

> Are there better cheaper mitigations for the "DatabaseNotReady at startup problem"

I don't think anything in this PR is a mitigation for the "DatabaseNotReady at startup problem". It's not changing when we populate the cache relative to other things that happen, just the mechanics of doing so.

> If we chose not to do this, we should probably record this in an issue as a potential optimization.

In turn, I have to admit that I'm not entirely sure what "this" refers to here - how are you proposing that the current implementation should change?

This conversations feels fruitful and like it should result in some docs improvements around the cache, which I'll try to take on as part of this PR...but I want to make sure I have an accurate mental model of your mental model of the system before doing so.

Reply:
James and I discussed this on slack and agreed to land the current behaviour and file an issue for potential future optimization.

Reply:
And here is that issue:  https://jira.mozilla.com/browse/SDK-189

Thanks, @rfk